design/StableRegions-20230209.md

```diff
@@ -0,0 +1,172 @@
+Stable Region Allocator Design for Motoko
+20230209
+Matthew Hammer
+
+## Purpose of this document
+
+This document gives a design for a new “stable region” allocator for Motoko’s runtime system, and accompanying `base` library.
+
+The role of this feature is summarized in the next section,
+followed by design considerations and details.
+
+
+## Role for "stable regions" in Motoko
+
+The current stable memory module in `base` has been "experimental" for a long time, and requires a more composable API to graduate from this status.
+
+Stable regions address the problem that today's `ExperimentalStableMemory` module only provides a single, monolithic memory that makes it unsuitable for directly building composable software parts.
+
+Stable regions permit a new API that supports composable use cases.
+
+Stable regions also bring Motoko closer to parity with Rust canister development support today, by giving a run-time-system-based analog of a special Rust library for stable data structures that allocates “pages” for them from stable memory in separate memory regions.
+
+
+## Design space
+
+The design space for the page allocator is defined by at least two
+tensions:
+
+ 1. fully-stable representation of allocator meta data **versus** fast load/store operations.
+
+ 2. Total scaling capacity **versus** minimum footprint for meta data.
+
+
+**Tension 1** is introduced because we want to avoid relying on the Motoko heap as the "ground truth" about the allocator's state.  If this heap is lost, as it is during an upgrade, then a developer may still want to recover all of the regions' data and meta data.
```

Comment:
>But yes, this does rely on the stable variables being reliable and wouldn't easily support recovery from a disaster without some additional info in stable memory (if recovery is a goal).

If we want to claim "parity" with the Rust version of this code, we should have the meta data in a form that "stays put", and is available during an upgrade failure, like with the preallocated tables.  Since that was important to @ielashi, and since I think that opinion is representative of our future devs, I took it as a design requirement.

Reply:
Yes, and I think that's a decent goal to have.